### PR TITLE
rgw: refactor boilerplate coroutine drain callers

### PR DESCRIPTION
**Pull request by yehudasa**
Add new helpers that make it easier to deal with concurrent coroutine spawn with a specific execution window. Use new helpers in the sync code.
Just moving code into macros was a bit tricky due to to the way `yield` works.
There are now two different `RGWCoroutine::drain_children()`. We could have a single method that handles both cases, but it seemed that the changes would be harder to validate and there is no real need for it currently.

Signed-off-by: Yehuda Sadeh <yehuda@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_coroutine.cc [hunk 1]
```diff
@@ -309,6 +323,12 @@ RGWCoroutinesStack *RGWCoroutinesStack::spawn(RGWCoroutine *op, bool wait)
   return spawn(NULL, op, wait);
 }
 
+RGWCoroutinesStack *RGWCoroutinesStack::prealloc_stack()
+{
+    preallocated_stack = env->manager->allocate_stack();
```
Comment (cbodley) [inline src/rgw/rgw_coroutine.cc:328]: if `preallocated_stack` is already allocated, it will leak here. consider storing it in a boost::intrusive_ptr

Comment (yehudasa) [reply src/rgw/rgw_coroutine.cc:328]: @cbodley I was considering to not allocate stack if preallocated stack existed, considering it a bug if prealloc_stack() is called twice without spawn. Can add that.

Comment (yehudasa) [reply src/rgw/rgw_coroutine.cc:329]: @cbodley pushed this change

**Comment by ivancich** [comment]
Teuthology runs:

- rgw/verify: http://pulpito.ceph.com/ivancich-2020-12-01_16:21:42-rgw-wip-eric-testing-1-distro-basic-smithi/
- rgw/(non-)verify: http://pulpito.ceph.com/ivancich-2020-12-01_16:21:59-rgw-wip-eric-testing-1-distro-basic-smithi/

File: src/rgw/rgw_data_sync.cc [hunk 2]
```diff
@@ -1726,18 +1743,13 @@ class RGWDataSyncShardCR : public RGWCoroutine {
...
-            while (collect(&ret, lease_stack.get())) {
-              if (ret < 0) {
-                tn->log(10, "a sync operation returned error");
-                /* we have reported this error */
-              }
-              /* not waiting for child here */
-            }
-          }
+
+          drain_all_but_stack_cb(lease_stack.get(),
```
Comment (cbodley) [inline src/rgw/rgw_data_sync.cc:1747]: yikes, data sync was respecting the `spawn_window` here before, but with this change it blocks after spawning each bucket sync

this PR added a `yield_spawn_window()` for bucket sync, so i opened https://github.com/ceph/ceph/pull/45567 to use that here in data sync too

**Comment by mattbenjamin** [comment]
@cbodley this is surprising, how did this go stale in 2020? :)

**Pull request merged**